test/io/pathfinder/test/ClusterSpec.java

```diff
@@ -0,0 +1,299 @@
+
+        return vehicle.apply(id, rand.nextDouble(), rand.nextDouble(), rand.nextInt());
+    }
+
+    public Commodity createCommodity() {
+        Commodity commodity = new Commodity();
+        Random rand = new Random();
+
+        commodity.startLatitude = rand.nextDouble();
+        commodity.endLatitude = rand.nextDouble();
+        commodity.startLongitude = rand.nextDouble();
+        commodity.endLongitude = rand.nextDouble();
+        commodity.param = rand.nextInt();
+
+        return commodity;
+    }
+
+    public Cluster createClusters() {
+        Cluster mainCluster = new Cluster();
+        Cluster cluster1 = new Cluster();
+        Cluster cluster2 = new Cluster();
+        Cluster cluster3 = new Cluster();
+        Cluster cluster4 = new Cluster();
+
+        Vehicle vehicle1 = createVehicle(0);
+        Vehicle vehicle2 = createVehicle(1);
+        Vehicle vehicle3 = createVehicle(2);
+
+        Commodity commodity1 = createCommodity();
+        Commodity commodity2 = createCommodity();
+        Commodity commodity3 = createCommodity();
+
+        cluster1.parent = mainCluster;
+        cluster4.parent = mainCluster;
+        cluster2.parent = cluster1;
+        cluster3.parent = cluster2;
+
+        mainCluster.subClusters.add(cluster1);
+        mainCluster.subClusters.add(cluster4);
+        cluster1.subClusters.add(cluster2);
+        cluster2.subClusters.add(cluster3);
+
+        mainCluster.vehicles.add(vehicle1);
+        cluster1.vehicles.add(vehicle2);
+        cluster2.vehicles.add(vehicle3);
+
+        mainCluster.commodities.add(commodity1);
+        cluster1.commodities.add(commodity2);
+        cluster2.commodities.add(commodity3);
+
+        return mainCluster;
+    }
+
+    @Test
+    public void ebeanModelShouldBeValid() {
+        Cluster mainCluster = createClusters();
+
+        mainCluster.save();
+
+        assertEquals(5, Cluster.find.all().size());
+        assertEquals(3, Vehicle.finder().all().size());
+        assertEquals(3, Commodity.find.all().size());
+    }
+
+    @Test
+    public void validPostShouldCreateClusterEasy() {
+        Helpers.running(fakeApp, () -> {
+            ObjectNode body = jsonNodeFactory.objectNode();
+
+            RequestBuilder request = new RequestBuilder()
+                    .bodyJson(body)
+                    .header("Content-Type", "application/json")
+                    .method(Helpers.POST)
+                    .uri("/cluster");
+
+            Result result = Helpers.route(request);
+
+            // Check for 'Created' Status Code
+            assertEquals(201, result.status());
+
+            ObjectNode resultJson = (ObjectNode) bodyForResult(result);
+
+            // Ensure that all fields were correctly written to the database
+            assertTrue("db record should have id", resultJson.hasNonNull("id"));
+
+        });
+    }
+
+    @Test
+    public void validPostShouldCreateClusterHard() {
+        Helpers.running(fakeApp, () -> {
```

Comment:
We might want to think about refactoring our unit tests before we have a ton of them. Thanks to play's excellent documentation, it was pretty tough to figure out the right way to set up the tests. I got it to work the way it is set up now, but there is definitely room for improvement.